pysatMadrigal/instruments/madrigal_pandas.py

```diff
@@ -0,0 +1,363 @@
+satellite in situ observations).
+
+Multiple kindat values are supported, as long as they are separated by commas.
+
+Although you can use this instrument module for any time-series data set, we
+highly recommend using the instrument-specific module if it exists.
+
+Please provide name (user) and email (password) when downloading data with this
+routine.
+
+Warnings
+--------
+All data downloaded under this general support is placed in the same directory,
+pysat_data_dir/madrigal/pandas/. For technical reasons, the file search
+algorithm for pysat's Madrigal support is set to permissive defaults. Thus, all
+instrument files downloaded via this interface will be picked up by the madrigal
+pandas pysat Instrument object unless the file_format keyword is used at
+instantiation.
+
+Files can be safely downloaded without knowing the file_format keyword,
+or equivalently, how Madrigal names the files. See `Examples` for more.
+
+Examples
+--------
+::
+
+
+    import datetime as dt
+    import pysat
+    import pysatMadrigal as py_mad
+
+    # Download DMSP data from Madrigal
+    dmsp_abi = pysat.Instrument(inst_module=py_mad.instruments.madrigal_pandas,
+                                tag='180', kindat='17110')
+    dmsp_abi.download(dt.datetime(2015, 1, 1), dt.datetime(2015, 1, 2),
+                      user='Firstname+Lastname', password='email@address.com')
+    dmsp_abi.load(date=dt.datetime(2015, 1, 1))
+
+"""
+
+import datetime as dt
+
+from pysat import logger
+
+from pysatMadrigal.instruments.methods import general
+
+# ----------------------------------------------------------------------------
+# Instrument attributes
+
+platform = 'madrigal'
+name = 'pandas'
+tags = dict()
+pandas_format = True
+
+# Assign only tags with pysat-compatible file format strings
+pandas_codes = general.known_madrigal_inst_codes(pandas_format=True)
+for tag in pandas_codes.keys():
+    try:
+        general.madrigal_file_format_str(tag, strict=True)
+        tags[tag] = pandas_codes[tag]
+    except ValueError:
+        pass
+
+inst_ids = {'': list(tags.keys())}  # There are too many kindat to track here
+
+# Local attributes
+#
+# Need a way to get the filename strings for a particular instrument unless
+# wildcards start working
+supported_tags = {ss: {tag: general.madrigal_file_format_str(tag)
+                       for tag in inst_ids[ss]} for ss in inst_ids.keys()}
+
+# ----------------------------------------------------------------------------
+# Instrument test attributes
+
+# Need to sort out test day setting for unit testing, maybe through a remote
+# function
+tag_dates = {'120': dt.datetime(1963, 11, 27), '170': dt.datetime(1998, 7, 1),
+             '180': dt.datetime(2000, 1, 1), '210': dt.datetime(1950, 1, 1),
+             '211': dt.datetime(1978, 1, 1), '212': dt.datetime(1957, 1, 1),
+             '7800': dt.datetime(2009, 11, 10), '8105': dt.datetime(2017, 9, 1)}
```

Comment:
```suggestion
             '7800': dt.datetime(2009, 11, 10), '8105': dt.datetime(2017, 9, 7)}
```

Reply:
There is data for Sept 7-11 on Madrigal. I can download data for Sept 7th, but to then load that data I have to load for Sept 1. The filename is `van_allen_2017_09.001.hdf5` which looks like it only has year and month in the filename..... 


```
In [19]: inst.download(
    ...:     dt.datetime(2017, 9, 7), user=Redacted, password=Redacted
    ...: )

In [20]: inst.load(date=dt.datetime(2017, 9, 7))
<ipython-input-20-18f7eb017f3c>:1: DeprecationWarning: Meta now contains a class for global metadata (MetaHeader). Default attachment of global attributes to Instrument will be Deprecated in pysat 3.2.0+. Set `use_header=True` in this load call or on Instrument instantiation to remove this warning.
  inst.load(date=dt.datetime(2017, 9, 7))

In [21]: inst.data
Out[21]: 
Empty DataFrame
Columns: []
Index: []

In [22]: inst.files.files
Out[22]: 
2017-09-01    van_allen_2017_09.001.hdf5
dtype: object

In [23]: inst.load(date=dt.datetime(2017, 9, 1))
pysat WARNING: The generalized Madrigal data Instrument can't support instrument-specific cleaning.
<ipython-input-23-146bd78005dd>:1: DeprecationWarning: Meta now contains a class for global metadata (MetaHeader). Default attachment of global attributes to Instrument will be Deprecated in pysat 3.2.0+. Set `use_header=True` in this load call or on Instrument instantiation to remove this warning.
  inst.load(date=dt.datetime(2017, 9, 1))

In [24]: inst.data
Out[24]: 
                     year  month  day  hour  min  sec  recno  kindat  kinst      ut1_unix      ut2_unix  ch_energy  lpeak_orb_el_flux  mid_grad_el_flux  plasmapause
2017-09-07 00:02:00  2017      9    7     0    2    0      0   10305   8105  1.504743e+09  1.504743e+09  2600000.0          11.563559               NaN          NaN
2017-09-07 04:54:00  2017      9    7     4   54    0      1   10305   8105  1.504760e+09  1.504760e+09  2600000.0          11.241525               NaN          NaN
2017-09-07 05:54:00  2017      9    7     5   54    0      2   10305   8105  1.504764e+09  1.504764e+09  2600000.0                NaN               NaN     4.690049
2017-09-07 09:05:00  2017      9    7     9    5    0      3   10305   8105  1.504775e+09  1.504775e+09  2600000.0          10.631356               NaN          NaN
2017-09-07 09:44:00  2017      9    7     9   44    0      4   10305   8105  1.504777e+09  1.504777e+09  2600000.0                NaN          4.037392          NaN
...                   ...    ...  ...   ...  ...  ...    ...     ...    ...           ...           ...        ...                ...               ...          ...
2017-09-11 15:40:00  2017      9   11    15   40    0     70   10305   8105  1.505144e+09  1.505144e+09  2600000.0                NaN               NaN     3.500927
2017-09-11 16:33:00  2017      9   11    16   33    0     71   10305   8105  1.505148e+09  1.505148e+09  2600000.0          12.190678               NaN          NaN
2017-09-11 18:05:00  2017      9   11    18    5    0     72   10305   8105  1.505153e+09  1.505153e+09  2600000.0                NaN               NaN     3.775340
2017-09-11 19:30:00  2017      9   11    19   30    0     73   10305   8105  1.505158e+09  1.505158e+09  2600000.0                NaN          2.992274          NaN
2017-09-11 20:02:00  2017      9   11    20    2    0     74   10305   8105  1.505160e+09  1.505160e+09  2600000.0          12.258475               NaN          NaN

[75 rows x 15 columns]

In [25]: 
```

Reply:
This looks like it'd need a special download routine to adjust the filename.  I guess we should remove this from the general instrument 🐕‍🦺

Reply:
Yeah.... there are also only 6 days total of data on the site.

Reply:
Not our highest priority 🐦